src/Catrobat/Security/FOSUBUserProvider.php

```diff
@@ -0,0 +1,121 @@
+    $property = $this->getProperty($response);
+    $username = $response->getUsername();
+    $service = $response->getResourceOwner()->getName();
+
+    $setter_name = 'set'.ucfirst($service);
+    $setter_id = $setter_name.'Id';
+    $setter_access_token = $setter_name.'AccessToken';
+    $access_token = $response->getAccessToken();
+
+    //Disconnect previous user
+    if (null !== $previousUser = $this->userManager->findUserBy([$property => $username]))
+    {
+      $previousUser->{$setter_id}(null);
+      $previousUser->{$setter_access_token}(null);
+      $this->userManager->updateUser($user);
+    }
+    //username is a unique integer
+    $user->{$setter_id}($username);
+    $user->{$setter_access_token}($access_token);
+    $this->userManager->updateUser($user);
+  }
+
+  public function loadUserByOAuthUserResponse(UserResponseInterface $response): UserInterface
+  {
+    $username = $response->getUsername();
+
+    $user = $this->userManager->findUserBy([$this->getProperty($response) => $username]);
+    $service = $response->getResourceOwner()->getName();
+    //make setter and getter names dynamic so we can use it for more services (Facebook, Google, Github)
+    $setter_name = 'set'.ucfirst($service);
+    $setter_id = $setter_name.'Id';
+    $setter_access_token = $setter_name.'AccessToken';
+    $access_token = $response->getAccessToken();
+    //register new user
+    if (null === $user)
+    {
+      $user = $this->userManager->findUserByEmail($response->getEmail());
+      //if user with the given email doesnt exists create a new user
+      if (null === $user)
+      {
+        /** @var User $user */
+        $user = $this->userManager->createUser();
+        //generate random username for example user12345678, needs to be discussed
+        $user->setUsername($this->createRandomUsername());
+        $user->setEmail($response->getEmail());
+        $user->setEnabled(true);
+        $user->setPassword($this->generateRandomPassword());
+        $user->setOauthUser(true);
+      }
+      $user->{$setter_id}($username);
+      $user->{$setter_access_token}($access_token);
+      $this->userManager->updateUser($user);
+
+      return $user;
+    }
+
+    //if user exists just proceed with hwioauthentication
+    $user = parent::loadUserByOAuthUserResponse($response);
+
+    //update access token
+    $user->{$setter_access_token}($access_token);
+
+    return $user;
+  }
+
+  private function generateRandomPassword(int $length = 32): string
+  {
+    $chars = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz'.
+      '0123456789-=~!@#$%&*()_+,.<>?;:[]{}|';
+
+    $pass = '';
+    $max = strlen($chars) - 1;
+
+    for ($i = 0; $i < $length; ++$i)
+    {
+      try
+      {
+        $pass .= $chars[random_int(0, $max)];
+      }
+      catch (Exception $e)
+      {
+      }
+    }
+
+    return $pass;
+  }
+
+  private function createRandomUsername(): string
+  {
+    $user = 'user';
+    $user_number = 1;
```

Comment:
similar to the reason we use uuids.

The management does not like this fixed incremented ids for various reasons.

On the other hand this process will be very slow once we have thousands of OAuth users.

It may does not look that pretty but we should try to use some kind of randomness in this naming scheme.

id = rand(0, 99999999999999)
hash_id = sha(id) 

something in this direction maybe? Still we should try to keep the usernames quite small, else most design aspects on the webpage look awful as soon as we show a username somewhere.

What's your view on this?  I will definetly ask UX for their opinion.

Reply:
well I will not be looking at this like an ID. Its different. But that it can be very slow thats where I agree. So we need to change this. We can maybe use crc32 for this hashing so that wouldnt be too long

Reply:
UX is asking:

can we use the fb/google/apple nickname somehow?
maybe just add (number in case of the same nickname)

Reply:
We can use first name and last name. Since in apple implemenation there is no nickname. But nickname is basically users full name so we can use that and then check if somebody already exists with that name and append number in that case